File: src/test/java/net/slipp/domain/tag/TagBuilder.java
package net.slipp.domain.tag;

import net.slipp.domain.user.SocialUser;

public class TagBuilder {
	private Long id;
	private String name;
	private int taggedCount;
	private boolean pooled;
	private Tag parentTag;
    private SocialUser owner;
    private String groupId;

	public static TagBuilder aTag() {
		return new TagBuilder();
	}
	
	public TagBuilder withId(Long id) {
		this.id = id;
		return this;
	}
	
	public TagBuilder withName(String name) {
		this.name = name;
		return this;
	}
	
	public TagBuilder withPooled(boolean pooled) {
		this.pooled = pooled;
		return this;
	}
	
	public TagBuilder withParentTag(Tag parentTag) {
		this.parentTag = parentTag;
		return this;
	}
	
	public TagBuilder withTaggedCount(int taggedCount) {
		this.taggedCount = taggedCount;
		return this;
	}
	
	public TagBuilder withOwner(SocialUser owner) {
	    this.owner = owner;
        return this;
    }
	
    public TagBuilder withGroupId(String groupId) {
        this.groupId = groupId;
        return this;
    }
	
	public Tag build() {
		TagInfo tagInfo = new TagInfo(owner, groupId, null);
		Tag tag = new Tag(name, parentTag, pooled, tagInfo) {
			public Long getTagId() {
				return id;
			}
		};
		
		for (int i = 0; i < taggedCount; i++) {
			tag.tagged();
		}
		
		return tag;
	}
}

File: src/test/scala/net/slipp/domain/tag/TagBuilder.scala
package net.slipp.domain.tag

import java.lang.Long

object TagBuilder {
  def aTag: TagBuilder = {
    return new TagBuilder
  }
}

class TagBuilder {
  private var id: Long = null
  private var name: String = null
  private var taggedCount: Int = 0
  private var parentTag: Tag = null
  private var groupId: String = null

  def withId(id: Long): TagBuilder = {
    this.id = id
    return this
  }

  def withName(name: String): TagBuilder = {
    this.name = name
    return this
  }

  def withParentTag(parentTag: Tag): TagBuilder = {
    this.parentTag = parentTag
    return this
  }

  def withTaggedCount(taggedCount: Int): TagBuilder = {
    this.taggedCount = taggedCount
    return this
  }

  def withGroupId(groupId: String): TagBuilder = {
    this.groupId = groupId
    return this
  }

  def build: Tag = {
    val tagInfo: TagInfo = new TagInfo(groupId, null)
    val tag: Tag = new Tag(name, parentTag, tagInfo) {
      override def getTagId: Long = {
        return id
      }
    }

    (0 to taggedCount).foreach(_ => tag.tagged())
    tag
  }
}
